torch/xpu/__init__.py

```diff
@@ -0,0 +1,203 @@
+r"""
+This package introduces support for the XPU backend, specifically tailored for
+Intel GPU optimization.
+
+You can use :func:`is_available()` to determine if your system supports XPU.
+"""
+from functools import lru_cache
+from typing import Any, Dict, Optional, Union
+
+import torch
+import torch._C
+from .. import device as _device
+from ._utils import _dummy_type, _get_device_index
+
+_device_t = Union[_device, str, int, None]
+
+
+def _is_compiled() -> bool:
+    r"""Return true if compile with XPU support."""
+    return torch._C._has_xpu
+
+
+if _is_compiled():
+    _XpuDeviceProperties = torch._C._XpuDeviceProperties
+    _exchange_device = torch._C._xpu_exchangeDevice
+    _maybe_exchange_device = torch._C._xpu_maybeExchangeDevice
+else:
+    # Define dummy if PyTorch was compiled without XPU
+    _XpuDeviceProperties = _dummy_type("_XpuDeviceProperties")  # type: ignore[assignment, misc]
+
+    def _exchange_device(device: int) -> int:
+        raise NotImplementedError("PyTorch was compiled without XPU support")
+
+    def _maybe_exchange_device(device: int) -> int:
+        raise NotImplementedError("PyTorch was compiled without XPU support")
+
+
+# TODO: Enable lazy init.
+if _is_compiled():
+    torch._C._xpu_init()
+
+
+@lru_cache(maxsize=1)
+def device_count() -> int:
+    r"""Return the number of XPU device available."""
+    if not _is_compiled():
+        return 0
+    return torch._C._xpu_getDeviceCount()
+
+
+def is_available() -> bool:
+    r"""Return a bool indicating if XPU is currently available."""
+    # This function nerver throws.
+    return device_count() > 0
+
+
+def is_bf16_supported():
+    r"""Return a bool indicating if the current XPU device supports dtype bfloat16."""
+    return True
```

Comment:
If all XPU devices support bf16, why this flag is needed?

Reply:
Currently, all XPU devices support bf16. We will refine this code when there is a new device breaks this rule. We hope this method can align to 'torch.cuda.is_bf16_supported`, reducing the effort to migrate the user's code.